File: src/controllers/review.js
import grpc from 'grpc'
import { grpcCode, RPCtoError } from "../configs/strings"
import { validationResult } from "express-validator";

var proto = grpc.load('./protos/book.proto')
var ReviewService = new proto.ReviewService(process.env.BOOK_SERVICE || "localhost:9010", grpc.credentials.createInsecure())
var BookService = new proto.BookService(process.env.BOOK_SERVICE || "localhost:9010", grpc.credentials.createInsecure())

proto = grpc.load('./protos/account.proto')
var AccountService = new proto.AccountService(process.env.ACCOUNT_SERVICE || "localhost:9000", grpc.credentials.createInsecure())

export class ReviewController {
    static insert (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        AccountService.GetById({id: req.userId}, (err, user) => {
            if (err)
                return res.json(RPCtoError(err));
            
            BookService.GetById({id: req.params.bookId}, (err, result) => {
                if(err)
                    return res.json(RPCtoError(err));
                
                let body_request = {
                    account_id: req.userId,
                    book_id: req.params.bookId,
                    user_name: user.name,
                    rating: req.body.rating,
                    resume: req.body.resume
                };
                
                ReviewService.Insert(body_request, (err, result) => {
                    if(err)
                        return res.json(RPCtoError(err));
                    
                    let response = JSON.parse(JSON.stringify(grpcCode.OK));
            
                    res.json(response);
                });
            });
        });
    }

    static getByAccountId (req, res) {
        ReviewService.GetByAccountId({id: req.userId}, (err, result) => {
            if(err)
                return res.json(RPCtoError(err));
            
            let response = JSON.parse(JSON.stringify(grpcCode.OK));
            response.body = result.reviews
    
            res.json(response);
        })
    }

    static getByBookId (req, res) {
        ReviewService.GetByBookId({id: req.params.bookId}, (err, result) => {
            if(err)
                return res.json(RPCtoError(err));
            
            let response = JSON.parse(JSON.stringify(grpcCode.OK));
            response.body = result.reviews
    
            res.json(response);
        })
    }

    static getById (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        ReviewService.GetById({id: req.params.reviewId}, (err, result) => {
            if(err)
                return res.json(RPCtoError(err));
            
            let response = JSON.parse(JSON.stringify(grpcCode.OK));
            response.body = result
    
            res.json(response);
        })
    }

    static delete (req, res) {
        const checker = validationResult(req);
        if (!checker.isEmpty()) {
            let result = grpcCode.InvalidArgument;
            result.errors = checker.array();

            return res.status(400).json(result);
        }

        ReviewService.Delete({id: req.params.reviewId, account_id: req.userId}, (err, result) => {
            if(err)
                return res.json(RPCtoError(err));
            
            let response = JSON.parse(JSON.stringify(grpcCode.OK));
    
            res.json(response);
        })
    }
}

File: services/review.go
package services

import (
	"context"
	"log"
	"ng-book-service/helpers"
	"ng-book-service/proto"
	"ng-book-service/repositories"
)

type ReviewService struct{ proto.ReviewServiceServer }

func (s *ReviewService) Get(ctx context.Context, in *proto.Empty) (*proto.Reviews, error) {
	log.Println("[*] [ReviewService] Get")

	reviews, err := repositories.Review.Get()
	if err != nil {
		return nil, err
	}

	result := proto.Reviews{}

	for _, obj := range *reviews {
		review, err := helpers.Review.FromModelToProto(&obj)

		if err != nil {
			return nil, err
		}

		result.Reviews = append(result.Reviews, review)
	}

	return &result, nil
}

func (s *ReviewService) GetByBookId(ctx context.Context, in *proto.Id) (*proto.Reviews, error) {
	log.Println("[*] [ReviewService] Get by book id")

	reviews, err := repositories.Review.GetByBookId(in.Id)
	if err != nil {
		return nil, err
	}

	result := proto.Reviews{}

	for _, obj := range *reviews {
		review, err := helpers.Review.FromModelToProto(&obj)

		if err != nil {
			return nil, err
		}

		result.Reviews = append(result.Reviews, review)
	}

	return &result, nil
}

func (s *ReviewService) GetByAccountId(ctx context.Context, in *proto.Id) (*proto.Reviews, error) {
	log.Println("[*] [ReviewService] Get by account id")

	reviews, err := repositories.Review.GetByAccountId(in.Id)
	if err != nil {
		return nil, err
	}

	result := proto.Reviews{}

	for _, obj := range *reviews {
		review, err := helpers.Review.FromModelToProto(&obj)

		if err != nil {
			return nil, err
		}

		result.Reviews = append(result.Reviews, review)
	}

	return &result, nil
}

func (s *ReviewService) GetById(ctx context.Context, in *proto.Id) (*proto.Review, error) {
	log.Println("[*] [ReviewService] GetById")

	review, err := repositories.Review.GetById(in.Id)
	if err != nil {
		return nil, err
	}

	result, err := helpers.Review.FromModelToProto(review)

	if err != nil {
		return nil, err
	}

	return result, nil
}

func (s *ReviewService) Insert(ctx context.Context, in *proto.Review) (*proto.Status, error) {
	log.Println("[*] [ReviewService] Insert")

	review, err := helpers.Review.FromProtoToModel(in)

	if err != nil {
		return nil, err
	}

	_, err = repositories.Review.Insert(*review)

	if err != nil {
		return nil, err
	}

	repositories.Review.RecalculateRating(review.BookId.String())

	return &proto.Status{Code: 0, Message: "Ok"}, nil
}

// func (s *ReviewService) Update(ctx context.Context, in *proto.Review) (*proto.Status, error) {
// 	log.Println("[*] [ReviewService] Update")

// 	review, err := helpers.Review.FromProtoToModel(in)

// 	if err != nil {
// 		return nil, err
// 	}

// 	_, err = repositories.Review.Update(*review)

// 	if err != nil {
// 		return nil, err
// 	}

// 	return &proto.Status{Code: 0, Message: "Ok"}, nil
// }

func (s *ReviewService) Delete(ctx context.Context, in *proto.IdAccountId) (*proto.Status, error) {
	log.Println("[*] [ReviewService] Delete")

	if _, err := repositories.Review.Delete(in.Id, in.AccountId); err != nil {
		return nil, err
	}

	return &proto.Status{Code: 0, Message: "OK"}, nil
}
